tests/core/data/layer/v1/sqlalchemy/test_repository.py

```diff
@@ -0,0 +1,193 @@
+import os
+import unittest
+from typing import Any, Dict
+
+import pandas as pd
+from sqlalchemy import Column, Integer, MetaData, String, Table, func, select
+from sqlalchemy.ext.asyncio import AsyncEngine, create_async_engine
+from sqlalchemy.sql import Select
+
+from py4ai.core.data.layer.v1.common.repository import (
+    QueryOptions,
+    SortingDirection,
+)
+from py4ai.core.data.layer.v1.sqlalchemy.criteria import (
+    SqlAlchemyQuery,
+    SqlAlchemySearchCriteria,
+)
+from py4ai.core.data.layer.v1.sqlalchemy.repository import SqlAlchemyRepository
+from py4ai.core.data.layer.v1.sqlalchemy.serializer import SqlAlchemySerializer
+from py4ai.core.utils.executors import AsyncExecutor
+from py4ai.core.utils.fs import create_dir_if_not_exists
+from tests import DATA_FOLDER, TMP_FOLDER
+from tests.core.data.layer.v1.base import (
+    CriteriaFactory,
+    Entity,
+    EntityDataSerializer,
+    EntityRepository,
+)
+
+
+class SqlCriteriaFactory(CriteriaFactory[SqlAlchemyQuery]):
+    def __init__(self, table: Table):
+        self.table = table
+
+    def by_cai(self, cai: int) -> SqlAlchemySearchCriteria:
+        return SqlAlchemySearchCriteria(self.table.c.cai == cai)
+
+    def from_birth_year(self, birth_year: int) -> SqlAlchemySearchCriteria:
+        return SqlAlchemySearchCriteria(self.table.c.birth_year >= birth_year)
+
+    def by_birth_year(self, birth_year: int) -> SqlAlchemySearchCriteria:
+        return SqlAlchemySearchCriteria(self.table.c.birth_year == birth_year)
+
+
+class SqlEntitySerializer(
+    SqlAlchemySerializer[int, int, Entity], EntityDataSerializer[int, Dict[str, Any]]
+):
+    def __init__(self, table: Table):
+        super(SqlEntitySerializer, self).__init__(table)
+
+    def to_object(self, entity: Entity) -> dict:
+        doc = entity.dict()
+        doc["id"] = self.to_object_key(self.get_key(entity))
+        return doc
+
+    def to_entity(self, document: dict) -> Entity:
+        return Entity(**document)
+
+    def to_object_key(self, key: int) -> int:
+        return key
+
+
+class SqlEntityRepository(
+    SqlAlchemyRepository[int, int, Entity],
+    EntityRepository[int, Dict[str, Any], SqlAlchemyQuery],
+):
+    def __init__(
+        self, engine: AsyncEngine, serializer: SqlAlchemySerializer[int, int, Entity]
+    ):
+        super().__init__(engine, serializer)
+        self.criteria = SqlCriteriaFactory(self.table)
+
+
+class TestRepository(unittest.TestCase):
```

Comment:
You should use `py4ai.core.tests.core.TestCase` instead of `unittest.TestCase`

Reply:
Done